### PR TITLE
Fixes #3122, #4600: Dark mode implementation - Home Screen, Recently Played Screen, Promotional Cards, Profile Progress, Ongoing Topic List, Completed Story List

### LINKED ISSUE #3122: Dark theme make colors of the images distorted
We currently do not support a dark theme but when a device enforces a dark theme on the app, then the colors of some images get distorted - 
Screenshot_2021-04-27-11-24-27-407_org oppia android

### LINKED ISSUE #4600: Doesn't have dark mode mock for promoted story cards
Need to add dark mode mock for `Promoted Story Cards`.

**To reproduce**
Once logged in, home activity comes under view and promoted story cards.

### PR DESCRIPTION
**Pull request by MohitGupta121**
## Explanation

Dark mode implementation - Home Screen, Recently Played Screen, Promotional Cards, Profile Progress, Ongoing Topic List, Completed Story List

Fixes #3122 : Dark theme make colors of the images distorted
Fixes #4600 : Doesn't have dark mode mock for promoted story  #cards

## Essential Checklist

- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [ ] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

### Completed Story List

   

If your PR includes UI-related changes, then:
- Add screenshots for portrait/landscape for both a tablet & phone of the before & after UI changes
- For the screenshots above, include both English and pseudo-localized (RTL) screenshots (see [RTL guide](https://github.com/oppia/oppia-android/wiki/RTL-Guidelines))
- Add a video showing the full UX flow with a screen reader enabled (see [accessibility guide](https://github.com/oppia/oppia-android/wiki/Accessibility-(A11y)-Guide))
- Add a screenshot demonstrating that you ran affected Espresso tests locally & that they're passing

### CODE REVIEW

File: app/src/main/res/drawable/interaction_read_only_white_background.xml
Comment (MohitGupta121) [inline app/src/main/res/drawable/interaction_read_only_white_background.xml]: @rt4914 this drawable is not use anywhere. So what to do with this?

File: app/src/main/res/layout-land/profile_progress_header.xml
Comment (MohitGupta121) [inline app/src/main/res/layout-land/profile_progress_header.xml:82]: As the `android:background="@drawable/grey_card_rounded_border"` this is not effecting any stroke/background. That's why I added this.

So we remove `android:background="@drawable/grey_card_rounded_border"` ?

File: app/src/main/res/layout-land/profile_progress_header.xml
Comment (MohitGupta121) [inline app/src/main/res/layout-land/profile_progress_header.xml:27]: This background is not effecting anything.

File: app/src/main/res/layout/recently_played_story_card.xml
Comment (MohitGupta121) [inline app/src/main/res/layout/recently_played_story_card.xml]: This style is keeping the text in bold stage but we not need bold text Story Name as per mock.

**Review by MohitGupta121 — COMMENTED** [review]
@rt4914 can you please review this.

File: app/src/main/res/drawable/interaction_read_only_white_background.xml
Comment (rt4914) [reply app/src/main/res/drawable/interaction_read_only_white_background.xml]: Simply remove this. If any tests fail that means we have to add it back.

Comment (MohitGupta121) [reply app/src/main/res/drawable/interaction_read_only_white_background.xml:7]: > Simply remove this. If any tests fail that means we have to add it back.

Okay, No test is failing as for now.

File: app/src/main/res/layout-land/profile_progress_header.xml
Comment (rt4914) [reply app/src/main/res/layout-land/profile_progress_header.xml:27]: I have cross checked it. You can remove it.

File: app/src/main/res/layout-land/profile_progress_header.xml
Comment (rt4914) [reply app/src/main/res/layout-land/profile_progress_header.xml:82]: Yes you can remove it.

File: app/src/main/res/layout/recently_played_story_card.xml
Comment (rt4914) [reply app/src/main/res/layout/recently_played_story_card.xml]: Although this style uses bold text we it is getting overridden in line 83 using `android:fontFamily="sans-serif-light"`

So I think we can keep this style.

File: app/src/main/res/drawable/interaction_read_only_white_background.xml
Comment (MohitGupta121) [reply app/src/main/res/drawable/interaction_read_only_white_background.xml]: Okay I'll check and remove it if all good.

File: app/src/main/res/layout-land/profile_progress_header.xml
Comment (MohitGupta121) [reply app/src/main/res/layout-land/profile_progress_header.xml:82]: Okay thanks.

File: app/src/main/res/layout-land/profile_progress_header.xml
Comment (MohitGupta121) [reply app/src/main/res/layout-land/profile_progress_header.xml:27]: > I have cross checked it. You can remove it.

Okay sure. Thanks.

File: app/src/main/res/layout/recently_played_story_card.xml
Comment (MohitGupta121) [reply app/src/main/res/layout/recently_played_story_card.xml]: I have to keep this style and remove font family or do something else with colors I think?

Comment (rt4914) [reply app/src/main/res/layout/recently_played_story_card.xml]: Keep
`style="@style/CaptionViewStart"`
as well as 
`android:fontFamily="sans-serif-light"`

Comment (MohitGupta121) [reply app/src/main/res/layout/recently_played_story_card.xml]: Okay

**Comment by MohitGupta121** [comment]
@rt4914 done with my review changes. 
Can you please TAL. Thanks

**Review by rt4914 — CHANGES_REQUESTED** [review]
@MohitGupta121 There is one issue with various cards.

In dark mode if you notice closely at the edges, you will see that there are white corners which are visible. They should not be visible.

For example in `prmoted_story_card` this is happening because we are using `app:roundedRectDrawableWithColor="@{viewModel.promotedStory.lessonThumbnail.backgroundColorRgb}"`
So basically to fix this in `promoted_story_card`, we first need to remove `app:roundedRectDrawableWithColor="@{viewModel.promotedStory.lessonThumbnail.backgroundColorRgb}"` and then in `PromotedStoryCardView` we should add `app:cardBackgroundColor="@android:color/transparent"` which should give correct results in both dark mode and light mode.

This above solution is for `promoted_story_card` only but we have various other cards too like `topics`, `ongoing_topic`, `completed_story`, etc so please focus on them too.

Once you have applied the changed and we are sure that it does look correct we will need Ben's approval too. 
If you wish to solve this in a separate PR do let me know and feel free to file an issue with correct screenshots describing the issue.

**Comment by MohitGupta121** [comment]
@rt4914 thanks for this review, I think we have to solve this issue in this PR it self as we have to complete the dark mode project.
But what you thinks? We file new issue or in same PR?

**Comment by rt4914** [comment]
> @rt4914 thanks for this review, I think we have to solve this issue in this PR it self as we have to complete the dark mode project. But what you thinks? We file new issue or in same PR?

@MohitGupta121 In that case, let me review this PR first, apply those changes first and in last commit make these corner-issue changes.

File: app/src/main/res/layout/completed_story_list_fragment.xml
Comment (rt4914) [inline app/src/main/res/layout/completed_story_list_fragment.xml:31]: `color_def_white` is getting used in `app:titleTextColor`.

Please update that.

File: app/src/main/res/layout/completed_story_item.xml
Comment (rt4914) [inline app/src/main/res/layout/completed_story_item.xml]: Instead of this use `component_color_shared_primary_text_color` color and also in document under `Post PR Changes` mention that we have remove `component_color_shared_content_item_text_view_color `

File: app/src/main/res/layout/ongoing_topic_list_fragment.xml
Comment (rt4914) [inline app/src/main/res/layout/ongoing_topic_list_fragment.xml:33]: `color_def_white` is getting used in `app:titleTextColor`. Please update that.

File: app/src/main/res/layout/ongoing_topic_item.xml
Comment (rt4914) [inline app/src/main/res/layout/ongoing_topic_item.xml]: Update the name of `component_color_shared_topic_card_primary_text_view_color` to `component_color_shared_topic_card_text_view_color`

Basically remove `primary` because we are actually using it in `Topic Title` (which is primary text) and also `Story Count` (which is secondary text).

File: app/src/main/res/layout/recently_played_activity.xml
Comment (rt4914) [inline app/src/main/res/layout/recently_played_activity.xml:28]: If you are using this style then make sure that in `styles.xml` for `ToolbarTextAppearance` you use the correct toolbar color. Currently its set to `color_def_white` which is incorrect. It should be `component_color_shared_activity_toolbar_text_color`

File: app/src/main/res/layout/section_title.xml
Comment (rt4914) [inline app/src/main/res/layout/section_title.xml]: Dont use `component_color_shared_content_item_text_view_color ` instead use `component_color_shared_primary_text_color`

File: app/src/main/res/layout/section_title.xml
Comment (rt4914) [inline app/src/main/res/layout/section_title.xml:33]: In line 23 we are using `      android:background="@color/divider"` 

Fix this please.

File: app/src/main/res/layout/recently_played_story_card.xml
Comment (rt4914) [inline app/src/main/res/layout/recently_played_story_card.xml]: Remove this and use `component_color_shared_primary_text_color`.

File: app/src/main/res/layout/profile_progress_fragment.xml
Comment (rt4914) [inline app/src/main/res/layout/profile_progress_fragment.xml]: This is not required. We can remove this.

File: app/src/main/res/layout/profile_progress_recently_played_story_card.xml
Comment (rt4914) [inline app/src/main/res/layout/profile_progress_recently_played_story_card.xml]: Use `component_color_shared_rounded_rect_background_color` instead.

File: app/src/main/res/layout/profile_progress_recently_played_story_card.xml
Comment (rt4914) [inline app/src/main/res/layout/profile_progress_recently_played_story_card.xml]: Use `component_color_shared_primary_text_color`

File: app/src/main/res/layout/profile_progress_recently_played_story_card.xml
Comment (rt4914) [inline app/src/main/res/layout/profile_progress_recently_played_story_card.xml]: Don't remove this style. It is required for RTL support.

File: app/src/main/res/layout/profile_progress_header.xml
Comment (rt4914) [inline app/src/main/res/layout/profile_progress_header.xml]: Remove `component_color_shared_content_item_text_view_color` and use already existing color as suggested earlier.

Comment (rt4914) [reply app/src/main/res/layout/profile_progress_header.xml]: All that to other textviews too in this file and to other similar header files.

File: app/src/main/res/layout/home_fragment.xml
Comment (rt4914) [inline app/src/main/res/layout/home_fragment.xml:14]: In `home_activity.xml` use `ToolbarTextAppearance` instead `ToolbarTitle` for `titleTextAppearance` so as to keep it consistent with other code.

File: app/src/main/res/layout/welcome.xml
Comment (rt4914) [inline app/src/main/res/layout/welcome.xml:34]: In `all_topics.xml` remove this `component_color_home_activity_select_a_topic_to_start_text_color` and `color_palette_select_a_topic_to_start_text_color` and 
use the `component_color_shared_primary_text_color`

File: app/src/main/res/layout/welcome.xml
Comment (rt4914) [inline app/src/main/res/layout/welcome.xml]: Remove this and use correct color `component_color_shared_content_item_text_view_color`

File: app/src/main/res/layout/welcome.xml
Comment (rt4914) [inline app/src/main/res/layout/welcome.xml]: Remove this and use correct color `component_color_shared_content_item_text_view_color`

File: app/src/main/res/layout/topic_summary_view.xml
Comment (rt4914) [inline app/src/main/res/layout/topic_summary_view.xml]: Remove this and use `component_color_shared_primary_text_color`

File: app/src/main/res/layout/topic_summary_view.xml
Comment (rt4914) [inline app/src/main/res/layout/topic_summary_view.xml]: Remove this and use `component_color_shared_primary_text_color`

File: app/src/main/res/layout/promoted_story_card.xml
Comment (rt4914) [inline app/src/main/res/layout/promoted_story_card.xml]: As suggested earlier remove this color and use `...shared_primary_text_color`

File: app/src/main/res/layout/promoted_story_card.xml
Comment (rt4914) [inline app/src/main/res/layout/promoted_story_card.xml]: As suggested earlier remove this color and use `...shared_primary_text_color`

File: app/src/main/res/layout/promoted_story_card.xml
Comment (rt4914) [inline app/src/main/res/layout/promoted_story_card.xml:63]: Instead of `bottom_rounded_rect_white_background` use `component_color_shared_rounded_rect_background_color`

File: app/src/main/res/layout/promoted_story_list.xml
Comment (rt4914) [inline app/src/main/res/layout/promoted_story_list.xml]: Remove this and use `...shared_primary_text_color`

**Review by rt4914 — CHANGES_REQUESTED** [review]
@MohitGupta121 I have suggested all changes. PTAL at those changes (don't regenerate screenshots right now we will do that after the last commit).

File: app/src/main/res/layout/recently_played_activity.xml
Comment (MohitGupta121) [reply app/src/main/res/layout/recently_played_activity.xml:28]: Okay that also we do for all colors in style. We have to document this also in Post PR changes.
I see you already added this thanks.

File: app/src/main/res/layout/section_title.xml
Comment (MohitGupta121) [reply app/src/main/res/layout/section_title.xml]: Okay I see everywhere I used this text, I replace all with `component_color_shared_primary_text_color `

File: app/src/main/res/layout/profile_progress_header.xml
Comment (MohitGupta121) [reply app/src/main/res/layout/profile_progress_header.xml]: Yes I fixed this for all others.

File: app/src/main/res/layout/topic_summary_view.xml
Comment (MohitGupta121) [reply app/src/main/res/layout/topic_summary_view.xml]: Actually here we have to use `component_color_shared_topic_card_text_view_color` as we need white color in both dark as well as light.

File: app/src/main/res/layout/promoted_story_card.xml
Comment (MohitGupta121) [reply app/src/main/res/layout/promoted_story_card.xml:63]: And can we delete `bottom_rounded_rect_white_background ` as it not use anywhere instead of this file.

File: app/src/main/res/layout/home_fragment.xml
Comment (MohitGupta121) [reply app/src/main/res/layout/home_fragment.xml:14]: So can we use `ToolbarTextAppearance ` in other file also such as `profile_progress_activity` also using the same `ToolbarTitle `.
Also there are other files also. So what you suggest @rt4914 ?

File: app/src/main/res/layout/section_title.xml
Comment (MohitGupta121) [reply app/src/main/res/layout/section_title.xml:33]: @rt4914 I fixed this add color `component_color_section_title_divider_color` can you suggest name and color as we don't have proper mock for this divider or use the shared one `component_color_shared_divider_color` ?

Also I found that stroke of the profile image set from the `circular_stroke_2dp_grey_32dp.xml` which is used in `ImageViewBindingAdapters.java` in line `100` and `103`. And also by `ImageViewBindingAdaptersTestActivity.kt` & `ImageViewBindingAdaptersTest.kt` So what to do with this Tests and Adapter file?

**Comment by MohitGupta121** [comment]
@rt4914 I commit the requested changes and leave comments on your review can you PTAL on that comments.

File: app/src/main/res/layout/ongoing_topic_list_fragment.xml
Comment (rt4914) [reply app/src/main/res/layout/ongoing_topic_list_fragment.xml:33]: Done.

File: app/src/main/res/layout/ongoing_topic_item.xml
Comment (rt4914) [reply app/src/main/res/layout/ongoing_topic_item.xml]: Done

File: app/src/main/res/layout/welcome.xml
Comment (rt4914) [reply app/src/main/res/layout/welcome.xml:34]: Done.

File: app/src/main/res/layout/welcome.xml
Comment (rt4914) [reply app/src/main/res/layout/welcome.xml]: Done.

File: app/src/main/res/layout/welcome.xml
Comment (rt4914) [reply app/src/main/res/layout/welcome.xml]: Done.

File: app/src/main/res/layout/recently_played_story_card.xml
Comment (rt4914) [reply app/src/main/res/layout/recently_played_story_card.xml]: Done.

File: app/src/main/res/layout/profile_progress_fragment.xml
Comment (rt4914) [reply app/src/main/res/layout/profile_progress_fragment.xml]: Done.

File: app/src/main/res/layout/home_fragment.xml
Comment (rt4914) [reply app/src/main/res/layout/home_fragment.xml:14]: Doing it for `profile_progress_activity ` and `home_activity ` is fine for now. For other activities in document mention that we have to remove `ToolbarTitle` post PR changes

File: app/src/main/res/layout/promoted_story_list.xml
Comment (rt4914) [reply app/src/main/res/layout/promoted_story_list.xml]: Done.

File: app/src/main/res/layout/promoted_story_card.xml
Comment (rt4914) [reply app/src/main/res/layout/promoted_story_card.xml:63]: Yes you can delete it if it is not getting used.

File: app/src/main/res/layout/promoted_story_card.xml
Comment (rt4914) [reply app/src/main/res/layout/promoted_story_card.xml]: Done.

File: app/src/main/res/layout/home_fragment.xml
Comment (MohitGupta121) [reply app/src/main/res/layout/home_fragment.xml:14]: Okay, I already done this for `profile_progress_activity `. 
And I'll add it to Post PR changes.

File: app/src/main/res/values/component_colors.xml
Comment (rt4914) [inline app/src/main/res/values/component_colors.xml]: Remove this comment and rename the variable to `component_color_recently_played_activity_section_title_divider_color `

File: app/src/main/res/layout/promoted_story_card.xml
Comment (MohitGupta121) [reply app/src/main/res/layout/promoted_story_card.xml:63]: okay sure.

File: app/src/main/res/layout/section_title.xml
Comment (rt4914) [reply app/src/main/res/layout/section_title.xml:33]: Use `component_color_shared_divider_color ` color for divider.

There is no Profile Image is this `section_title.xml` file. I am unable to understand what are you referring to? If this is from some other file then please reference that in a separate comment.

**Comment by rt4914** [comment]
@MohitGupta121 I have replied to all the comments.

@MohitGupta121 Also, you had not replied to all the comments earlier. For now I have replied as "Done" and resolved them but please make sure that you reply to all comments otherwise your PR will get blocked on "Unresolved conversations".

**Comment by MohitGupta121** [comment]
> @MohitGupta121 I have replied to all the comments.
> 
> @MohitGupta121 Also, you had not replied to all the comments earlier. For now I have replied as "Done" and resolved them but please make sure that you reply to all comments otherwise your PR will get blocked on "Unresolved conversations".

Okay sure, I take care of this next time.
But I not have to make it as resolved?

File: app/src/main/res/values/component_colors.xml
Comment (MohitGupta121) [reply app/src/main/res/values/component_colors.xml]: Actually there is no need of this color if we using `component_color_shared_divider_color` as per this [comment](https://github.com/oppia/oppia-android/pull/4786#discussion_r1050786884)

File: app/src/main/res/layout/section_title.xml
Comment (MohitGupta121) [reply app/src/main/res/layout/section_title.xml:33]: Done.

File: app/src/main/res/layout/promoted_story_card.xml
Comment (MohitGupta121) [reply app/src/main/res/layout/promoted_story_card.xml:63]: Done.

**Comment by MohitGupta121** [comment]
@rt4914 I done with the lastest requested changes. Can you PTAL.

@rt4914 In this [comment](https://github.com/oppia/oppia-android/pull/4786#discussion_r1050786884) I taking about the stroke image is set from the logical files. So my question is we have to set dark mode colors in Test file `ImageViewBindingAdaptersTest.kt` also?
And the stroke of the profile image set from the `circular_stroke_2dp_grey_32dp.xml` which is used in `ImageViewBindingAdapters.java` in line `100` and `103`.

File: app/src/main/res/values/component_colors.xml
Comment (MohitGupta121) [reply app/src/main/res/values/component_colors.xml]: Done.

**Comment by rt4914** [comment]
> circular_stroke_2dp_grey_32dp

1. We don't need to apply dark-mode changes to Test files. Its completely optional.
2. In `circular_stroke_2dp_grey_32dp ` there multiple colors which are not supporting dark-mode and also are not from `component_colors.xml` files so this needs to be fixed.
3. We should always updated update adapter files like `ImageViewBindingAdapter` to support dark mode and correct color variables.
4. Most importantly the line (100-103) and file that you are mentioning, ie.e  `circular_stroke_2dp_grey_32dp` it getting used in `StoryActivity` so it should not be part of this PR.

So I am just confused why are you asking question related to `StoryActivity` in this PR which has nothing to do with `StoryActivity`. 

Am I missing something else here?

**Comment by MohitGupta121** [comment]
> Am I missing something else here?

Okay I just see that file that's why I have confused is this not a part of this PR.
Now it's clear to me, I have to done this fix `circular_stroke_2dp_grey_32dp` in the part of StoryActivity?

**Comment by rt4914** [comment]
> > Am I missing something else here?
> 
> Okay I just see that file that's why I have confused is this not a part of this PR. Now it's clear to me, I have to done this fix `circular_stroke_2dp_grey_32dp` in the part of StoryActivity?

Yes that's correct.

**Review by rt4914 — APPROVED** [review]
LGTM, Thanks.

**Comment by MohitGupta121** [comment]
> In dark mode if you notice closely at the edges, you will see that there are white corners which are visible. They should not be visible.

@rt4914 So now I start the corners fixes? As if everything is fine in this PR. And if we are in the last stage of this PR?

**Pull request merged**
